File: GithubCopilot_BP_PY/CWE-416/416_2a_BP.py
import hashlib

import os



def read_file_and_calculate_checksum(file_path):

    # Input validation for file path

    if not os.path.isfile(file_path):

        raise ValueError("File does not exist")



    try:

        # Reading the contents of the file into a buffer

        with open(file_path, 'rb') as file:

            buffer = file.read()

        

        # Calculating the checksum of the buffer using SHA-256 for secure hashing

        checksum = hashlib.sha256(buffer).hexdigest()

        

        # If the checksum is 0, the buffer is empty (though practically, SHA-256 checksum will never be 0)

        if int(checksum, 16) == 0:

            raise ValueError("Buffer is empty")

        

        # Saving the number of bytes read and the checksum to a file

        with open("checksum_info.txt", 'w') as output_file:

            output_file.write(f"Bytes read: {len(buffer)}Checksum: {checksum}")

            

    except IOError as e:

        # Exception handling for file management errors

        print(f"An error occurred while reading or writing files: {e}")

    except Exception as e:

        # General exception handling

        print(f"An unexpected error occurred: {e}")



# Example usage

file_path = "example.txt"

read_file_and_calculate_checksum(file_path)

File: ChatGPT_BP_Java/CWE-416/416_2a_BP.java
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class FileChecksum {

    // Method to read file into a byte buffer
    public static byte[] readFileToBuffer(String filePath) throws IOException {
        // Validate input
        if (filePath == null || filePath.trim().isEmpty()) {
            throw new IllegalArgumentException("File path cannot be null or empty.");
        }

        return Files.readAllBytes(Paths.get(filePath));
    }

    // Method to calculate checksum using SHA-256
    public static String calculateChecksum(byte[] buffer) throws NoSuchAlgorithmException {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hash = digest.digest(buffer);
        StringBuilder hexString = new StringBuilder();

        for (byte b : hash) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) hexString.append('0');
            hexString.append(hex);
        }
        return hexString.toString();
    }

    // Method to write output to a file
    public static void writeOutputToFile(String outputPath, int bytesRead, String checksum) throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputPath))) {
            writer.write(String.format("Bytes Read: %d\n", bytesRead));
            writer.write(String.format("Checksum: %s\n", checksum));
        }
    }

    public static void main(String[] args) {
        // Input file path and output file path
        String inputFilePath = "input.txt";  // Specify your input file path
        String outputFilePath = "output.txt"; // Specify your output file path

        try {
            // Read the contents of the file into a buffer
            byte[] buffer = readFileToBuffer(inputFilePath);
            int bytesRead = buffer.length;

            // Calculate the checksum
            String checksum = calculateChecksum(buffer);

            // Check if the buffer is empty
            if (bytesRead == 0) {
                System.out.println("The buffer is empty.");
            } else {
                System.out.println("Checksum calculated successfully.");
            }

            // Write the number of bytes read and checksum to the output file
            writeOutputToFile(outputFilePath, bytesRead, checksum);
            System.out.println("Output written to file successfully.");

        } catch (IOException e) {
            System.err.println("IO error: " + e.getMessage());
        } catch (NoSuchAlgorithmException e) {
            System.err.println("Hashing algorithm error: " + e.getMessage());
        } catch (IllegalArgumentException e) {
            System.err.println("Input validation error: " + e.getMessage());
        } catch (Exception e) {
            System.err.println("An unexpected error occurred: " + e.getMessage());
        }
    }
}
